src/Bicep.Core/Emit/ExpressionConverter.cs

```diff
@@ -115,30 +113,27 @@ public LanguageExpression ConvertExpression(SyntaxBase expression)
 
         public ExpressionConverter CreateConverterForIndexReplacement(SyntaxBase nameSyntax, SyntaxBase? indexExpression, SyntaxBase newContext)
         {
-            // local function
-            SyntaxBase GetArrayExpression(LocalVariableSymbol localVariable)
-            {
-                var parent = this.context.SemanticModel.Binder.GetParent(localVariable.DeclaringLocalVariable);
-                return parent switch
-                {
-                    ForSyntax @for => @for.Expression,
-                    _ => throw new NotImplementedException($"Local variable has an unexpected parent of type '{parent?.GetType().Name}")
-                };
-            }
-
             var inaccessibleLocals = this.context.DataFlowAnalyzer.GetInaccessibleLocalsAfterSyntaxMove(nameSyntax, newContext);
-            switch (inaccessibleLocals.Count)
+            var inaccessibleLocalLoops = inaccessibleLocals.Select(local => GetEnclosingForExpression(local)).Distinct().ToList();
```

Comment:
>inaccessibleLocalLoops [](start = 16, length = 22)

Most of the changes in this file are to remove code duplication between the paths involved in indexed collection access and normal locals replacement.